File: Search/Search_a_2D_Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int top = 0;
        int rowLastIndex = matrix.length - 1;
        int colLastIndex = matrix[0].length - 1;
        int bot = rowLastIndex;
        int curRow = 0;

        while (top <= bot) {
            curRow = (top + bot) / 2;
            if (target < matrix[curRow][0]) {
                bot = curRow - 1;
            } else if (target > matrix[curRow][colLastIndex]) {
                top = curRow + 1;
            } else {
                break;
            }
        }

        int left = 0, right = colLastIndex;

        while (left <= right) {
            int mid = (right + left) / 2;
            if (matrix[curRow][mid] < target) {
                left = mid + 1;
            } else if (matrix[curRow][mid] > target) {
                right = mid - 1;
            } else {
                return true;
            }
        }
        
        return false;
    }
}

File: Search/Search_a_2D_Matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        top, bot = 0, len(matrix) - 1

        while top <= bot:
            curRow = (top + bot) // 2
            if target < matrix[curRow][0]:
                bot = curRow - 1
            elif target > matrix[curRow][-1]:
                top = curRow + 1
            else:
                break

        left, right = 0, len(matrix[0]) - 1

        while left <= right:
            mid = (right + left) // 2
            if matrix[curRow][mid] < target:
                left = mid + 1
            elif matrix[curRow][mid] > target:
                right = mid - 1
            else: 
                return True

        return False